### PR TITLE
Nic coalesce support

### PR DESCRIPTION
**Pull request by daandemeyer**
Includes one commit from @yuwata that still needs to land, will rebase once that happens so don't merge yet.

I'm not entirely sure about the naming of the options. I went with names that mostly match the names from the fields of the corresponding ethtool ioctl struct with two modifications:

- Make sure config names for microseconds always end in USec
- Add Coalesce to fields that didn't contain Coalesce in the name already to clearly indicate that they are coalescing settings.

Implements one of the requested features in #20392

### CODE REVIEW

File: src/shared/ethtool-util.c [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #include "socket-util.h"
 #include "string-table.h"
 #include "strxcpyx.h"
+#include "strv.h"
```
Comment (poettering) [inline src/shared/ethtool-util.c:17]: please keep this sorted

File: src/shared/ethtool-util.c [hunk 2]
```diff
@@ -1101,3 +1102,169 @@ int config_parse_wol(
...
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata
+) {
```
Comment (poettering) [inline src/shared/ethtool-util.c]: please move to end of previous line

File: src/shared/ethtool-util.c [hunk 3]
```diff
@@ -1101,3 +1102,169 @@ int config_parse_wol(
...
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse coalesce setting value, ignoring: %s", rvalue);
+                return 0;
+        }
+
+        if (STR_IN_SET(lvalue, "StatsBlockCoalesceUSec", "CoalesceRateSampleInterval") && k < 1) {
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                           "Invalid %s= value, ignoring: %s", lvalue, rvalue);
+                return 0;
+        }
```
Comment (poettering) [inline src/shared/ethtool-util.c:1177]: please use separate parser functions if the fields are sufficeintly different. also use parse_sec() for parsing time values.

StatsBlockCoalesceUSec should probably be called StatisticsBlockCoalesceSec=. i.e, please no unnecessary abbreviations. And we use seconds as default time unit everywhere, please do so here too. If people specify other time units explicitly that takes precedence of course, and parse_sec() takes care of that.

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/shared/ethtool-util.c [hunk 4]
```diff
@@ -1101,3 +1102,201 @@ int config_parse_wol(
...
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+        u32_opt *dst = data;
+        uint32_t k;
+        int r;
+
+        dst->value = 0;
+        dst->set = false;
```
Comment (yuwata) [inline src/shared/ethtool-util.c]: Please do not update `dst`, as `rvalue` may be an invalid string.

File: src/shared/ethtool-util.c [hunk 5]
```diff
@@ -1101,3 +1102,201 @@ int config_parse_wol(
...
+
+        if (isempty(rvalue))
+                return 0;
```
Comment (yuwata) [inline src/shared/ethtool-util.c]: Please unset the values in that case:
```
if (isempty(rvalue)) {
    dst->value = 0;
    dst->set = false;
    return 0;
}
```

File: src/shared/ethtool-util.c [hunk 6]
```diff
@@ -1101,3 +1102,201 @@ int config_parse_wol(
...
+                void *userdata) {
+        usec_opt *dst = data;
+        usec_t usec;
+        int r;
+
+        dst->value = 0;
+        dst->set = false;
+
+        if (isempty(rvalue))
+                return 0;
```
Comment (yuwata) [inline src/shared/ethtool-util.c]: Same here.

File: src/shared/ethtool-util.h [hunk 7]
```diff
@@ -62,6 +62,11 @@ typedef struct u32_opt {
         bool set;
 } u32_opt;
 
+typedef struct usec_opt {
+        uint64_t value;
```
Comment (yuwata) [inline src/shared/ethtool-util.h]: `uint64_t` -> `usec_t`

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/shared/ethtool-util.c [hunk 8]
```diff
@@ -1101,3 +1102,201 @@ int config_parse_wol(
...
+                return -errno;
+
+        if (coalesce->use_adaptive_rx_coalesce >= 0)
+                UPDATE(ecmd.use_adaptive_rx_coalesce, (uint32_t) coalesce->use_adaptive_rx_coalesce, need_update);
+
+        if (coalesce->use_adaptive_tx_coalesce >= 0)
+                UPDATE(ecmd.use_adaptive_tx_coalesce, (uint32_t) coalesce->use_adaptive_tx_coalesce, need_update);
+
+        if (coalesce->rx_coalesce_usecs.set)
+                UPDATE(ecmd.rx_coalesce_usecs, coalesce->rx_coalesce_usecs.value, need_update);
```
Comment (daandemeyer) [inline src/shared/ethtool-util.c:1240]: Here (and for all the other fields containing usecs) we convert from unsigned 64-bit to unsigned 32-bit implicitly. Do we need to check anywhere if the value provided by the user is too large to fit inside unsigned 32-bit?

**Comment by DaanDeMeyer** [comment]
Removed usec_opt and just reuse u32_opt for the usec values instead. In the `coalesce_parse_sec()` function we check if the parsed microseconds fit in the unsigned 32-bit value field of u32_opt.

**Review by yuwata — APPROVED** [review]
LGTM.

**Pull request merged**